scripts/test/Muon/loadfile_presenter_test.py

```diff
@@ -0,0 +1,510 @@
+import six
+import sys
+import unittest
+
+if sys.version_info.major == 3:
+    from unittest import mock
+else:
+    import mock
+
+from PyQt4.QtGui import QApplication
+
+from Muon.GUI.MuonAnalysis.loadfile.load_file_view import BrowseFileWidgetView
+from Muon.GUI.MuonAnalysis.loadfile.load_file_presenter import BrowseFileWidgetPresenter
+from Muon.GUI.MuonAnalysis.loadfile.load_file_model import BrowseFileWidgetModel
+from Muon.GUI.Common.muon_load_data import MuonLoadData
+
+
+class IteratorWithException:
+    """Wraps a simple iterable (i.e. list) so that it throws a ValueError on a particular index."""
+
+    def __init__(self, iterable, throw_on_index):
+        self.max = len(iterable)
+        self.iterable = iter(iterable)
+
+        self.throw_indices = [index for index in throw_on_index if index < self.max]
+
+    def __iter__(self):
+        self.n = 0
+        return self
+
+    def __next__(self):
+
+        if self.n in self.throw_indices:
+            next(self.iterable)
+            self.n += 1
+            raise ValueError()
+        elif self.n == self.max:
+            raise StopIteration()
+        else:
+            self.n += 1
+            return next(self.iterable)
+
+    next = __next__
+
+
+# global QApplication (get errors if > 1 instance in the code)
+QT_APP = QApplication([])
+
+
+class LoadFileWidgetPresenterTest(unittest.TestCase):
+    def run_test_with_and_without_threading(f):
+
+        def run_twice(self):
+            f(self)
+            self.setUp()
+            self.presenter._use_threading = False
+            f(self)
+
+        return run_twice
+
+    class Runner:
+
+        def __init__(self, thread):
+            self.QT_APP = QT_APP
```

Comment:
what is this for?

Reply:
This keeps a local reference to the global instance of the Qt EventLoop, for use within the runner. Removing this line does interfere with the tests, possibly because tests are threaded (although I'm not sure about this)

Reply:
what about the runner?

Reply:
The runner is necessary for any tests which involve threaded code. The thread requires a main event loop to be running when it starts; otherwise no signal/slot connections are made, and none of its log messages make it through to the console.

The `Runner` class comes from some Qt forum discussion and is the neatest way of achieving this I have found.